### ISSUE TITLE
Track QC results for each QC test

### ISSUE THREAD

**Issue by ggalibert** [aodn/imos-toolbox#492]
From @CraigSteinberg:
> Tracking which QC test failed a particular test will allow better expert QC analysis and the possibility of automated qc deciding on whether or not a particular file needs further effort.

I suppose you mean "which data value failed which QC test".

At the moment, the toolbox enables you to see which proportion of data have been flagged by each test in the "QC stats" window:
![screenshot from 2018-03-14 14-47-41](https://user-images.githubusercontent.com/3498686/37382389-b7f8cb90-2796-11e8-880f-061b30164562.png)

This information is also present in the output NetCDF file in the global attribute ```quality_control_log```. For example:
```
imosImpossibleDateQC(dateMin=01/01/2007, dateMax=21/04/2017) did not fail on any TIME sample.
imosImpossibleLocationSetQC(distanceKmPlusMinusThreshold=2.5) did not fail on any LATITUDE sample.
imosImpossibleLocationSetQC(distanceKmPlusMinusThreshold=2.5) did not fail on any LONGITUDE sample.
imosInOutWaterQC(in=20/06/11 12:55:00, out=31/08/11 13:30:00) flagged 1601 TEMP samples with flag Bad_data.
imosInOutWaterQC(in=20/06/11 12:55:00, out=31/08/11 13:30:00) flagged 1601 PRES samples with flag Bad_data.
imosInOutWaterQC(in=20/06/11 12:55:00, out=31/08/11 13:30:00) flagged 1601 DEPTH samples with flag Bad_data.
imosGlobalRangeQC(min=-2.5, max=40) did not fail on any TEMP sample.
imosGlobalRangeQC(min=-5, max=12000) did not fail on any PRES sample.
imosGlobalRangeQC(min=-5, max=12000) did not fail on any DEPTH sample.
imosImpossibleDepthQC(zNominalMargin=15, maxAngle=70 => min=99.7923, max=136.0703) flagged 120 PRES samples with flag Bad_data.
imosImpossibleDepthQC(zNominalMargin=15, maxAngle=70 => min=89, max=125) flagged 120 DEPTH samples with flag Bad_data.
```
It doesn't give you the detail of which sample data fails which test but at least this is a first help for further effort / expert QC analysis. You can tell when an automated test has gone nuts.

If you wish to go further and **track the QC result for each QC test at every sample data** then 2 solutions that can be considered. One or the other, or both, could be adopted:
1. this information is accessible/visible via the toolbox (stored in memory).
2. this information is written in the output FV01 NetCDF file.

Note that solution 2, to my knowledge, has never been adopted by any data centre. Argo considered it but then rejected it due to the complexity of the produced NetCDF that would confuse users. You have to bear in mind that as of today, there are 18 QC tests for timeseries so that for each physical variable sampled by the instrument there would potentially be 1 + 18 + 1 = 20 variables in the NetCDF file while a WQM samples on average 15 physical variable -> 15 x 20 = 300!

**Comment by CraigSteinberg** [comment]
Yes sorry - the request was a bit unclear. OK that helps me understand.
I am just thinking that when we do any major reprocessing it would mean we could target just the data affected by a particular QC test fail. Something the community can discuss the merits of further.

It would be useful for the toolbox to load a FV01 netCDF file for manual editing and possible further partial processing. @sspagnol  has written a parser is beleive. Not sure if it could be applied here.

**Comment by ggalibert** [comment]
The toolbox can already read a FV01 NetCDF file so we can imagine further editing / processing. However the same limitations as in discussion https://github.com/aodn/imos-toolbox/issues/491 would apply.

**Comment by ggalibert** [comment]
Another approach would be to store multiple boolean information in a single integer variable: http://blog.millermedeiros.com/using-integers-to-store-multiple-boolean-values/ .
Could complement the current flag system since more complex.

**Comment by mhidas** [comment]
CF already has a way to describe such combined boolean variables, using variable attributes `flag_masks` and `flag_meanings`: http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#flags (see Example 3.4)

**Comment by ggalibert** [comment]
There are currently 19 QC tests for time series so we would need to store this multiple boolean information into a uint32 (could host up to 32 QC test results) which uses as much memory as a single precision float (like most IMOS parameters) although might be compressed a lot better. 

Note that not all QC tests have binary outcomes though, [some](https://github.com/aodn/imos-toolbox/wiki/QCProcedures#adcp-tilt-test---imostiltvelocitysetqc---compulsory) can deliver "good", "probably_good" or "bad"...

**Comment by mhidas** [comment]
> Note that not all QC tests have binary outcomes though, some can deliver "good", "probably_good" or "bad"...

That does complicate things a little, but you could just assign two bits to those tests (if not too many...).

**Comment by ocehugo** [comment]
2 cents: If you got 20 tests and the limit is 32, I recommend thinking again in compressing them.
I say that because sooner or later new tests will appears and you will hit a wall and have to reimplment the code/logic again...

You could avoid that by creating superQC flags (groups): combinatorics of some of them. Say "qcImpossible" is the combination of impossibledates/locations/etc. this can also grow again, but it may be a remarkable achievement.

Nonetheless, if this is going to be used only be operators, why not a new file (FV_01 auxiliary) to assist with results? In the other case...n4 groups (again!?).
